## The Electronic Nose Dataset for Plant Oil with Different Compositions

In [33]:
import os
import pandas as pd
from pathlib import Path

# Define the dataset path
dataset_path = Path('/Users/momthu/Documents/duhoc/Hungary/elte/research/oil/dataset_Oils')

# Get all oil type folders
oil_folders = sorted([d for d in dataset_path.iterdir() if d.is_dir()])
print(f"Found {len(oil_folders)} oil type folders:")
for folder in oil_folders:
    print(f"  - {folder.name}")


Found 19 oil type folders:
  - CO
  - COO
  - COO1
  - COO2
  - COO3
  - COO4
  - COO5
  - RO0
  - RO1
  - RO2
  - RO3
  - RO4
  - RO5
  - SO0
  - SO1
  - SO2
  - SO3
  - SO4
  - SO5


In [34]:
# Check columns consistency across all files
columns_dict = {}
total_files = 0
files_with_issues = []

for oil_folder in oil_folders:
    txt_files = sorted(oil_folder.glob('*.txt'))
    print(f"\n{oil_folder.name}: {len(txt_files)} files")
    
    if txt_files:
        # Read first file to get columns
        first_file = txt_files[0]
        df = pd.read_csv(first_file)
        columns = tuple(df.columns)
        
        if columns not in columns_dict:
            columns_dict[columns] = []
        
        columns_dict[columns].append((oil_folder.name, first_file.name, len(txt_files)))
        print(f"  Columns: {list(df.columns)}")
        print(f"  Shape: {df.shape}")
        total_files += len(txt_files)
        
        # Check if all files in this folder have the same columns
        for txt_file in txt_files[1:]:
            df_check = pd.read_csv(txt_file)
            if tuple(df_check.columns) != columns:
                files_with_issues.append((oil_folder.name, txt_file.name))

print(f"\n{'='*60}")
print(f"TOTAL FILES: {total_files}")
print(f"{'='*60}")


CO: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO1: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO2: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO3: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO4: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO5: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

RO0: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

RO1: 70 fil

In [35]:
# Summary of columns consistency
print("\n" + "="*70)
print("SUMMARY OF COLUMN CONSISTENCY ANALYSIS")
print("="*70)

if len(columns_dict) == 1:
    print("\n✓ ALL FILES HAVE CONSISTENT COLUMNS!")
    print(f"\nTotal unique column structure(s): {len(columns_dict)}")
    
    # Display the single column structure
    for columns, oil_info in columns_dict.items():
        print(f"\nColumn names ({len(columns)} columns):")
        for i, col in enumerate(columns, 1):
            print(f"  {i}. {col}")
        
        print(f"\nFiles using this structure ({len(oil_info)} oil types):")
        total_files_with_structure = 0
        for oil_type, first_file, num_files in oil_info:
            print(f"  - {oil_type}: {num_files} files")
            total_files_with_structure += num_files
        print(f"\n  Total files: {total_files_with_structure}")
else:
    print(f"\n✗ FOUND {len(columns_dict)} DIFFERENT COLUMN STRUCTURES!")
    print("\nColumn structures:")
    for i, (columns, oil_info) in enumerate(columns_dict.items(), 1):
        print(f"\n  Structure {i}: {columns}")
        print(f"  Used in: {[info[0] for info in oil_info]}")

if files_with_issues:
    print(f"\n⚠ WARNING: Found {len(files_with_issues)} files with inconsistent columns")
    print("Files with issues:")
    for oil_type, filename in files_with_issues[:10]:  # Show first 10
        print(f"  - {oil_type}/{filename}")
    if len(files_with_issues) > 10:
        print(f"  ... and {len(files_with_issues) - 10} more")
else:
    print("\n✓ All files within each oil type have consistent columns")

print("\n" + "="*70)


SUMMARY OF COLUMN CONSISTENCY ANALYSIS

✓ ALL FILES HAVE CONSISTENT COLUMNS!

Total unique column structure(s): 1

Column names (7 columns):
  1. Time [s]
  2. TGS2620
  3. TGS2602
  4. TGS2603
  5. TGS2610
  6. TGS2611
  7. TGS2600

Files using this structure (19 oil types):
  - CO: 70 files
  - COO: 70 files
  - COO1: 70 files
  - COO2: 70 files
  - COO3: 70 files
  - COO4: 70 files
  - COO5: 70 files
  - RO0: 70 files
  - RO1: 70 files
  - RO2: 70 files
  - RO3: 70 files
  - RO4: 70 files
  - RO5: 70 files
  - SO0: 70 files
  - SO1: 70 files
  - SO2: 70 files
  - SO3: 70 files
  - SO4: 70 files
  - SO5: 70 files

  Total files: 1330

✓ All files within each oil type have consistent columns



In [36]:
# Additional details about data structure
print("\nDATA STRUCTURE DETAILS:")
print("="*70)

# Check data shape consistency
print("\nChecking data shape consistency...")
shape_dict = {}

for oil_folder in oil_folders:
    txt_files = sorted(oil_folder.glob('*.txt'))
    if txt_files:
        df_sample = pd.read_csv(txt_files[0])
        shape = df_sample.shape
        
        if shape not in shape_dict:
            shape_dict[shape] = []
        shape_dict[shape].append(oil_folder.name)

print(f"\nFound {len(shape_dict)} unique data shape(s):")
for shape, oil_types in sorted(shape_dict.items()):
    print(f"  Shape {shape}: {len(oil_types)} oil types")
    if len(oil_types) <= 5:
        print(f"    Oil types: {', '.join(oil_types)}")

# Data type information
print("\nData type information:")
sample_df = pd.read_csv(list(oil_folders[0].glob('*.txt'))[0])
print(sample_df.dtypes)

print("\nSample data (first 5 rows from CO_1.txt):")
print(sample_df.head())


DATA STRUCTURE DETAILS:

Checking data shape consistency...

Found 1 unique data shape(s):
  Shape (150, 7): 19 oil types

Data type information:
Time [s]      int64
TGS2620     float64
TGS2602     float64
TGS2603     float64
TGS2610     float64
TGS2611     float64
TGS2600     float64
dtype: object

Sample data (first 5 rows from CO_1.txt):
   Time [s]   TGS2620   TGS2602   TGS2603   TGS2610   TGS2611   TGS2600
0         1  0.044517  0.030207  0.029896  0.008595  0.018460  0.033513
1         2  0.044510  0.030220  0.029909  0.008594  0.018462  0.033507
2         3  0.044500  0.030234  0.029908  0.008592  0.018458  0.033500
3         4  0.044494  0.030246  0.029910  0.008594  0.018458  0.033490
4         5  0.044485  0.030262  0.029915  0.008593  0.018457  0.033486


## Analysis Results

### Key Findings ✓

1. **Column Standardization: FULLY STANDARDIZED**
   - All 1,330 data files across all 19 oil types use identical columns
   - No inconsistencies found

2. **Column Structure**
   - **7 columns total**: 1 time column + 6 sensor columns
   - Time column: `Time [s]` (integer values, 1-150 seconds)
   - Sensor columns: `TGS2620`, `TGS2602`, `TGS2603`, `TGS2610`, `TGS2611`, `TGS2600` (all float values)

3. **Data Shape Consistency**
   - All files have identical dimensions: **(150 rows, 7 columns)**
   - Each file contains 150 time measurements

4. **Oil Types**
   - 19 different oil types represented
   - Each type has exactly 70 data files
   - Oil types: CO, COO, COO1-5, RO0-5, SO0-5

### Sensor Information
The dataset uses 6 sensors from the TGS series (electronic nose sensors):
- TGS2620: TVOC & smoke detection
- TGS2602: Odor & smoke detection  
- TGS2603: Odor detection
- TGS2610: Alcohol detection
- TGS2611: General air quality
- TGS2600: General air quality

In [37]:
# Statistics about sample counts in each file
print("\n" + "="*70)
print("SAMPLE COUNT STATISTICS")
print("="*70)

file_stats = []

for oil_folder in oil_folders:
    txt_files = sorted(oil_folder.glob('*.txt'))
    
    if txt_files:
        for txt_file in txt_files:
            df = pd.read_csv(txt_file)
            sample_count = len(df)
            file_stats.append({
                'Oil Type': oil_folder.name,
                'File': txt_file.name,
                'Samples': sample_count,
                'Path': str(txt_file)
            })

# Create dataframe for easier analysis
stats_df = pd.DataFrame(file_stats)

# Summary statistics
print("\nOVERALL STATISTICS:")
print(f"  Total files: {len(stats_df)}")
print(f"  Total samples across all files: {stats_df['Samples'].sum():,}")
print(f"  Average samples per file: {stats_df['Samples'].mean():.0f}")
print(f"  Min samples in a file: {stats_df['Samples'].min()}")
print(f"  Max samples in a file: {stats_df['Samples'].max()}")

# By oil type
print("\nSAMPLES BY OIL TYPE:")
oil_type_stats = stats_df.groupby('Oil Type').agg({
    'Samples': ['count', 'sum', 'mean', 'min', 'max']
}).round(0)

for oil_type in sorted(stats_df['Oil Type'].unique()):
    subset = stats_df[stats_df['Oil Type'] == oil_type]
    print(f"\n  {oil_type}:")
    print(f"    Files: {len(subset)}")
    print(f"    Total samples: {subset['Samples'].sum():,}")
    print(f"    Samples per file: {subset['Samples'].iloc[0]} (all files identical)")

# Check if all files have same number of samples
unique_samples = stats_df['Samples'].unique()
if len(unique_samples) == 1:
    print(f"\n✓ ALL FILES HAVE SAME NUMBER OF SAMPLES: {unique_samples[0]} samples per file")
else:
    print(f"\n⚠ Files have DIFFERENT sample counts:")
    for sample_count in sorted(unique_samples):
        count = len(stats_df[stats_df['Samples'] == sample_count])
        print(f"  {sample_count} samples: {count} files")

print("\n" + "="*70)


SAMPLE COUNT STATISTICS

OVERALL STATISTICS:
  Total files: 1330
  Total samples across all files: 199,500
  Average samples per file: 150
  Min samples in a file: 150
  Max samples in a file: 150

SAMPLES BY OIL TYPE:

  CO:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO1:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO2:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO3:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO4:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO5:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  RO0:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  RO1:


## Summary Statistics

### Sample Count Analysis ✓

**Hoàn toàn đồng nhất về số lượng mẫu!**

- **Tổng số file**: 1,330 files
- **Tổng số mẫu**: 199,500 samples
- **Mẫu/file**: **150 mẫu** (tất cả file đều giống nhau)
- **Mẫu/loại dầu**: 10,500 mẫu (70 files × 150 mẫu)

### Phân bố dữ liệu
- **CO**: 70 files × 150 samples = 10,500 samples
- **COO + COO1-5**: 7 types × 70 files × 150 = 73,500 samples  
- **RO0-5**: 6 types × 70 files × 150 = 63,000 samples
- **SO0-5**: 6 types × 70 files × 150 = 63,000 samples

### Kết luận
✓ Dữ liệu hoàn toàn chuẩn hóa:
- Cột: Giống nhau (7 cột)
- Kích thước: Giống nhau (150 hàng × 7 cột)
- Số file: Đồng đều (70 files/loại dầu)
- Tổng mẫu: 199,500 samples

## Data Cleaning and Train/Validation/Test Split for TimeGAN

Mục tiêu:
- Đọc toàn bộ file `.txt` thành chuỗi thời gian 3D `(n_sequences, seq_len, n_features)`.
- Clean dữ liệu: ép kiểu số, xử lý thiếu dữ liệu, xử lý trùng thời gian, đảm bảo độ dài chuỗi cố định.
- Chia dữ liệu thành 3 tập `train/validation/test` theo tỷ lệ 70/15/15, giữ phân bố theo từng loại dầu.
- Chuẩn hóa MinMax theo **train-only** để tránh data leakage.
- Lưu output ra thư mục `processed_timegan/` để dùng trực tiếp cho TimeGAN.

In [38]:
import numpy as np

# Configuration for TimeGAN prep
RANDOM_SEED = 42
TRAIN_RATIO = 0.70
VAL_RATIO = 0.15
TEST_RATIO = 0.15

FEATURE_COLUMNS = ['TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
TIME_COLUMN = 'Time [s]'
EXPECTED_SEQ_LEN = 150

output_dir = dataset_path.parent / 'processed_timegan'
output_dir.mkdir(exist_ok=True)

rng = np.random.default_rng(RANDOM_SEED)

print('Configuration ready.')
print(f'Output directory: {output_dir}')
print(f'Feature columns ({len(FEATURE_COLUMNS)}): {FEATURE_COLUMNS}')

Configuration ready.
Output directory: /Users/momthu/Documents/duhoc/Hungary/elte/research/oil/processed_timegan
Feature columns (6): ['TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']


In [39]:
# Build metadata of all files
records = []

for oil_folder in sorted([d for d in dataset_path.iterdir() if d.is_dir()]):
    for txt_file in sorted(oil_folder.glob('*.txt')):
        records.append({
            'oil_type': oil_folder.name,
            'file_name': txt_file.name,
            'file_path': str(txt_file)
        })

meta_df = pd.DataFrame(records)

print(f'Total sequences: {len(meta_df)}')
print(f'Oil types: {meta_df["oil_type"].nunique()}')
print('\nSequences per oil type:')
print(meta_df['oil_type'].value_counts().sort_index())

Total sequences: 1330
Oil types: 19

Sequences per oil type:
oil_type
CO      70
COO     70
COO1    70
COO2    70
COO3    70
COO4    70
COO5    70
RO0     70
RO1     70
RO2     70
RO3     70
RO4     70
RO5     70
SO0     70
SO1     70
SO2     70
SO3     70
SO4     70
SO5     70
Name: count, dtype: int64


In [40]:
# Clean one sequence file and return fixed-length feature matrix (seq_len, n_features)
def clean_sequence_file(file_path: str,
                        feature_cols=FEATURE_COLUMNS,
                        time_col=TIME_COLUMN,
                        expected_len=EXPECTED_SEQ_LEN):
    df_raw = pd.read_csv(file_path)

    # Keep only relevant columns and coerce to numeric.
    df = df_raw[[time_col] + feature_cols].copy()
    for col in [time_col] + feature_cols:
        df[col] = pd.to_numeric(df[col], errors='coerce')

    # Sort by time and collapse duplicated timestamps by mean.
    df = df.sort_values(time_col)
    df = df.groupby(time_col, as_index=False)[feature_cols].mean()

    # Reindex to expected timeline and interpolate missing values.
    expected_time = np.arange(1, expected_len + 1)
    df = df.set_index(time_col).reindex(expected_time)
    df[feature_cols] = df[feature_cols].interpolate(method='linear', limit_direction='both')
    df[feature_cols] = df[feature_cols].ffill().bfill()

    # Last fallback for pathological cases.
    if df[feature_cols].isna().any().any():
        df[feature_cols] = df[feature_cols].fillna(0.0)

    return df[feature_cols].to_numpy(dtype=np.float32)


# Build 3D tensor and track cleaning issues
X_list = []
issues = []

for i, row in meta_df.iterrows():
    try:
        x_seq = clean_sequence_file(row['file_path'])
        if x_seq.shape != (EXPECTED_SEQ_LEN, len(FEATURE_COLUMNS)):
            issues.append((row['file_name'], f'unexpected shape {x_seq.shape}'))
        X_list.append(x_seq)
    except Exception as e:
        issues.append((row['file_name'], str(e)))
        X_list.append(np.zeros((EXPECTED_SEQ_LEN, len(FEATURE_COLUMNS)), dtype=np.float32))

X = np.stack(X_list, axis=0)
y = meta_df['oil_type'].to_numpy()

print(f'X shape for TimeGAN: {X.shape} (n_sequences, seq_len, n_features)')
print(f'y shape: {y.shape}')
print(f'Cleaning issues found: {len(issues)}')
if issues:
    print('First 5 issues:')
    for item in issues[:5]:
        print(' -', item)

X shape for TimeGAN: (1330, 150, 6) (n_sequences, seq_len, n_features)
y shape: (1330,)
Cleaning issues found: 0


In [41]:
# Stratified split by oil type (70/15/15)
train_idx, val_idx, test_idx = [], [], []

for oil_type, grp in meta_df.groupby('oil_type'):
    idx = grp.index.to_numpy(copy=True)
    rng.shuffle(idx)

    n = len(idx)
    n_train = int(round(n * TRAIN_RATIO))
    n_val = int(round(n * VAL_RATIO))

    # Ensure all samples are assigned.
    if n_train + n_val > n:
        n_val = n - n_train
    n_test = n - n_train - n_val

    train_idx.extend(idx[:n_train])
    val_idx.extend(idx[n_train:n_train + n_val])
    test_idx.extend(idx[n_train + n_val:])

train_idx = np.array(sorted(train_idx))
val_idx = np.array(sorted(val_idx))
test_idx = np.array(sorted(test_idx))

X_train, y_train = X[train_idx], y[train_idx]
X_val, y_val = X[val_idx], y[val_idx]
X_test, y_test = X[test_idx], y[test_idx]

print('Split sizes:')
print(f'  Train: {X_train.shape[0]} sequences')
print(f'  Val:   {X_val.shape[0]} sequences')
print(f'  Test:  {X_test.shape[0]} sequences')
print(f'  Total: {X_train.shape[0] + X_val.shape[0] + X_test.shape[0]} sequences')

print('\nClass balance per split (counts):')
print('Train:\n', pd.Series(y_train).value_counts().sort_index())
print('\nVal:\n', pd.Series(y_val).value_counts().sort_index())
print('\nTest:\n', pd.Series(y_test).value_counts().sort_index())

Split sizes:
  Train: 931 sequences
  Val:   190 sequences
  Test:  209 sequences
  Total: 1330 sequences

Class balance per split (counts):
Train:
 CO      49
COO     49
COO1    49
COO2    49
COO3    49
COO4    49
COO5    49
RO0     49
RO1     49
RO2     49
RO3     49
RO4     49
RO5     49
SO0     49
SO1     49
SO2     49
SO3     49
SO4     49
SO5     49
Name: count, dtype: int64

Val:
 CO      10
COO     10
COO1    10
COO2    10
COO3    10
COO4    10
COO5    10
RO0     10
RO1     10
RO2     10
RO3     10
RO4     10
RO5     10
SO0     10
SO1     10
SO2     10
SO3     10
SO4     10
SO5     10
Name: count, dtype: int64

Test:
 CO      11
COO     11
COO1    11
COO2    11
COO3    11
COO4    11
COO5    11
RO0     11
RO1     11
RO2     11
RO3     11
RO4     11
RO5     11
SO0     11
SO1     11
SO2     11
SO3     11
SO4     11
SO5     11
Name: count, dtype: int64


In [42]:
# MinMax normalization using train-only statistics (avoid leakage)
train_2d = X_train.reshape(-1, X_train.shape[-1])
feat_min = train_2d.min(axis=0)
feat_max = train_2d.max(axis=0)
feat_range = np.where((feat_max - feat_min) == 0, 1.0, (feat_max - feat_min))

def minmax_scale(x, fmin, frange):
    return (x - fmin) / frange

X_train_scaled = minmax_scale(X_train, feat_min, feat_range).astype(np.float32)
X_val_scaled = minmax_scale(X_val, feat_min, feat_range).astype(np.float32)
X_test_scaled = minmax_scale(X_test, feat_min, feat_range).astype(np.float32)

# Keep data in [0, 1] for TimeGAN stability.
X_train_scaled = np.clip(X_train_scaled, 0.0, 1.0)
X_val_scaled = np.clip(X_val_scaled, 0.0, 1.0)
X_test_scaled = np.clip(X_test_scaled, 0.0, 1.0)

print('Scaled tensor ranges:')
print(f'  Train min/max: {X_train_scaled.min():.4f} / {X_train_scaled.max():.4f}')
print(f'  Val   min/max: {X_val_scaled.min():.4f} / {X_val_scaled.max():.4f}')
print(f'  Test  min/max: {X_test_scaled.min():.4f} / {X_test_scaled.max():.4f}')

Scaled tensor ranges:
  Train min/max: 0.0000 / 1.0000
  Val   min/max: 0.0000 / 0.8481
  Test  min/max: 0.0006 / 1.0000


## Experimental Protocol: Variable Dataset Sizes with Multi-Seed Training

Protocol:
1. Fix test set at 20% of data
2. Split remaining 80% into variable training sizes (tăng dần)
3. For each training size: train 4-5 models with different random seeds
4. Scaling experiment: train with 3 samples → generate 4-50 samples → evaluate improvement
5. Progressive scaling: increase from 3 → 4 → ... → 10-15 sample seeds

Mục tiêu: Tìm optimal dataset size cho quality/diversity trade-off


In [ ]:
import torch
import torch.nn as nn
from torch.optim import Adam
import matplotlib.pyplot as plt
from scipy import stats
from sklearn.metrics import roc_auc_score, accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.decomposition import PCA
import json
from tqdm import tqdm
import time

# Setup device
device = torch.device('cpu')  # Using CPU for stability
print(f'Device: {device}')

# Create output directories for experiments
experiment_dir = output_dir / 'experiments_multiseed'
experiment_dir.mkdir(exist_ok=True)

print(f'Experiment directory: {experiment_dir}')


In [ ]:
# STEP 1: Create fixed TEST set (20% of data)
# Remaining 80% will be used for variable training sizes

TEST_RATIO_FIXED = 0.20
SEEDS_PER_SIZE = 4  # Number of seeds per training size
TRAIN_SIZES = [50, 100, 200, 400, 600, 931]  # Variable training sizes

# Stratified split to create fixed test set
test_idx_fixed = []
trainval_idx = []

for oil_type, grp in meta_df.groupby('oil_type'):
    idx = grp.index.to_numpy(copy=True)
    rng.shuffle(idx)
    
    n = len(idx)
    n_test = int(round(n * TEST_RATIO_FIXED))
    
    test_idx_fixed.extend(idx[:n_test])
    trainval_idx.extend(idx[n_test:])

test_idx_fixed = np.array(sorted(test_idx_fixed))
trainval_idx = np.array(sorted(trainval_idx))

# Fixed test set (scaled)
X_test_fixed, y_test_fixed = X[test_idx_fixed], y[test_idx_fixed]
X_test_fixed_scaled = minmax_scale(X_test_fixed, feat_min, feat_range).astype(np.float32)
X_test_fixed_scaled = np.clip(X_test_fixed_scaled, 0.0, 1.0)

# Remaining data for training (will be resampled at various sizes)
X_trainval = X[trainval_idx]
y_trainval = y[trainval_idx]

print('Fixed Test Set Created:')
print(f'  Test size: {len(test_idx_fixed)} sequences (20%)')
print(f'  Train+Val available: {len(trainval_idx)} sequences (80%)')
print(f'  Test set class distribution:')
print(pd.Series(y_test_fixed).value_counts().sort_index())

# Save indices for reproducibility
np.save(experiment_dir / 'test_idx_fixed.npy', test_idx_fixed)
np.save(experiment_dir / 'trainval_idx.npy', trainval_idx)
np.save(experiment_dir / 'X_test_fixed_scaled.npy', X_test_fixed_scaled)
np.save(experiment_dir / 'y_test_fixed.npy', y_test_fixed)
print(f'\nFixed test set saved to {experiment_dir}')


In [ ]:
# STEP 2: Function for stratified sampling of training sets

def stratified_sample(X, y, n_samples, random_seed=42):
    """
    Sample n_samples from X, y maintaining class balance (stratified sampling)
    """
    rng_local = np.random.default_rng(random_seed)
    
    sampled_idx = []
    
    for oil_type in np.unique(y):
        # Find all indices for this oil type
        mask = (y == oil_type)
        type_idx = np.where(mask)[0]
        
        # Calculate how many samples from this class
        class_count = np.sum(mask)
        total_count = len(y)
        class_fraction = class_count / total_count
        n_from_class = max(1, int(round(n_samples * class_fraction)))
        
        # Sample from this class
        if n_from_class >= len(type_idx):
            sampled_idx.extend(type_idx)
        else:
            sampled_idx.extend(rng_local.choice(type_idx, size=n_from_class, replace=False))
    
    sampled_idx = np.array(sorted(sampled_idx))
    return sampled_idx[:n_samples]  # Trim to exact size if needed


# Test the function
test_sample_idx = stratified_sample(X_trainval, y_trainval, 100, random_seed=42)
print(f'Test sampling function:')
print(f'  Requested: 100 samples')
print(f'  Got: {len(test_sample_idx)} samples')
print(f'  Class distribution:')
print(pd.Series(y_trainval[test_sample_idx]).value_counts().sort_index())


In [ ]:
# STEP 3: TimeGAN Architecture (same as before)

class RecurrentBlock(nn.Module):
    def __init__(self, hidden_dim, num_layers=2):
        super().__init__()
        self.rnn = nn.LSTM(hidden_dim, hidden_dim, num_layers=num_layers, batch_first=True)
    
    def forward(self, x):
        out, _ = self.rnn(x)
        return out


class Embedder(nn.Module):
    def __init__(self, n_features, hidden_dim):
        super().__init__()
        self.fc1 = nn.Linear(n_features, hidden_dim)
        self.rnn = RecurrentBlock(hidden_dim, num_layers=2)
        self.fc2 = nn.Linear(hidden_dim, hidden_dim)
    
    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.rnn(x)
        x = torch.relu(self.fc2(x))
        return x


class Recovery(nn.Module):
    def __init__(self, hidden_dim, n_features):
        super().__init__()
        self.fc1 = nn.Linear(hidden_dim, hidden_dim)
        self.rnn = RecurrentBlock(hidden_dim, num_layers=2)
        self.fc2 = nn.Linear(hidden_dim, n_features)
    
    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.rnn(x)
        x = self.fc2(x)
        return x


class Supervisor(nn.Module):
    def __init__(self, hidden_dim):
        super().__init__()
        self.fc1 = nn.Linear(hidden_dim, hidden_dim)
        self.rnn = RecurrentBlock(hidden_dim, num_layers=2)
        self.fc2 = nn.Linear(hidden_dim, hidden_dim)
    
    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.rnn(x)
        x = torch.relu(self.fc2(x))
        return x


class Generator(nn.Module):
    def __init__(self, hidden_dim):
        super().__init__()
        self.fc1 = nn.Linear(hidden_dim, hidden_dim)
        self.rnn = RecurrentBlock(hidden_dim, num_layers=2)
        self.fc2 = nn.Linear(hidden_dim, hidden_dim)
    
    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.rnn(x)
        x = torch.relu(self.fc2(x))
        return x


class Discriminator(nn.Module):
    def __init__(self, hidden_dim):
        super().__init__()
        self.fc1 = nn.Linear(hidden_dim, hidden_dim)
        self.rnn = RecurrentBlock(hidden_dim, num_layers=2)
        self.fc2 = nn.Linear(hidden_dim, 1)
    
    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.rnn(x)
        x = self.fc2(x)
        return x


print('TimeGAN architecture classes defined')


In [ ]:
# STEP 4: Training function for TimeGAN

def train_timegan(X_train, hidden_dim=48, epochs_e=8, epochs_s=8, epochs_g=20, batch_size=32, seed=42):
    """
    Train TimeGAN on X_train (should be scaled to [0, 1])
    Returns: trained models
    """
    torch.manual_seed(seed)
    np.random.seed(seed)
    
    n_samples, seq_len, n_features = X_train.shape
    
    # Initialize models
    embedder = Embedder(n_features, hidden_dim).to(device)
    recovery = Recovery(hidden_dim, n_features).to(device)
    supervisor = Supervisor(hidden_dim).to(device)
    generator = Generator(hidden_dim).to(device)
    discriminator = Discriminator(hidden_dim).to(device)
    
    # Optimizers
    opt_e = Adam(embedder.parameters(), lr=0.001)
    opt_r = Adam(recovery.parameters(), lr=0.001)
    opt_s = Adam(supervisor.parameters(), lr=0.001)
    opt_g = Adam(generator.parameters(), lr=0.001)
    opt_d = Adam(discriminator.parameters(), lr=0.001)
    
    # Loss functions
    mse_loss = nn.MSELoss()
    bce_loss = nn.BCEWithLogitsLoss()
    
    X_train_tensor = torch.from_numpy(X_train).to(device)
    
    # PHASE 1: Embedder + Recovery
    print(f'  Phase 1: Embedder + Recovery training ({epochs_e} epochs)')
    for epoch in range(epochs_e):
        epoch_loss = 0
        for i in range(0, n_samples, batch_size):
            batch = X_train_tensor[i:i+batch_size]
            
            opt_e.zero_grad()
            opt_r.zero_grad()
            
            h = embedder(batch)
            x_rec = recovery(h)
            loss = mse_loss(x_rec, batch)
            
            loss.backward()
            opt_e.step()
            opt_r.step()
            
            epoch_loss += loss.item()
        
        if (epoch + 1) % 2 == 0 or epoch == 0:
            print(f'    Epoch {epoch+1}/{epochs_e}, Loss: {epoch_loss/max(1, n_samples//batch_size):.6f}')
    
    # PHASE 2: Supervisor
    print(f'  Phase 2: Supervisor training ({epochs_s} epochs)')
    for epoch in range(epochs_s):
        epoch_loss = 0
        for i in range(0, n_samples, batch_size):
            batch = X_train_tensor[i:i+batch_size]
            
            opt_s.zero_grad()
            
            h = embedder(batch).detach()
            h_pred = supervisor(h)
            loss = mse_loss(h_pred[:, :-1, :], h[:, 1:, :])
            
            loss.backward()
            opt_s.step()
            
            epoch_loss += loss.item()
        
        if (epoch + 1) % 2 == 0 or epoch == 0:
            print(f'    Epoch {epoch+1}/{epochs_s}, Loss: {epoch_loss/max(1, n_samples//batch_size):.6f}')
    
    # PHASE 3: Joint training (Generator + Discriminator)
    print(f'  Phase 3: Joint training ({epochs_g} epochs)')
    for epoch in range(epochs_g):
        g_loss_total = 0
        d_loss_total = 0
        
        for i in range(0, n_samples, batch_size):
            batch = X_train_tensor[i:i+batch_size]
            batch_size_actual = batch.shape[0]
            
            # Generate synthetic data
            z = torch.randn(batch_size_actual, seq_len, hidden_dim).to(device)
            
            # Generator step
            opt_g.zero_grad()
            h_fake = generator(z)
            x_fake = recovery(h_fake)
            
            # Try to fool discriminator
            d_real = discriminator(embedder(batch).detach())
            d_fake = discriminator(h_fake.detach())
            
            g_loss = bce_loss(d_fake, torch.ones_like(d_fake))
            
            g_loss.backward()
            opt_g.step()
            g_loss_total += g_loss.item()
            
            # Discriminator step
            opt_d.zero_grad()
            
            h_fake = generator(z).detach()
            d_real = discriminator(embedder(batch).detach())
            d_fake = discriminator(h_fake)
            
            d_loss = bce_loss(d_real, torch.ones_like(d_real)) + bce_loss(d_fake, torch.zeros_like(d_fake))
            
            d_loss.backward()
            opt_d.step()
            d_loss_total += d_loss.item()
        
        if (epoch + 1) % 5 == 0 or epoch == 0:
            print(f'    Epoch {epoch+1}/{epochs_g}, G_Loss: {g_loss_total/max(1, n_samples//batch_size):.6f}, D_Loss: {d_loss_total/max(1, n_samples//batch_size):.6f}')
    
    return embedder, recovery, supervisor, generator, discriminator


print('Training function defined')


In [ ]:
# STEP 5: Generation and Evaluation Functions

def generate_synthetic_data(generator, recovery, n_samples, seq_len, hidden_dim, device='cpu'):
    """Generate n_samples of synthetic sequences"""
    generator.eval()
    recovery.eval()
    
    with torch.no_grad():
        z = torch.randn(n_samples, seq_len, hidden_dim).to(device)
        h_fake = generator(z)
        x_fake = recovery(h_fake)
        x_fake = x_fake.cpu().numpy()
    
    return x_fake.astype(np.float32)


def evaluate_synthetic_quality(X_real, X_fake):
    """
    Compute quality metrics comparing real vs fake data
    Returns: dict with metrics
    """
    metrics = {}
    
    # 1. Mean and Std difference
    real_mean = X_real.mean(axis=(0, 1))
    fake_mean = X_fake.mean(axis=(0, 1))
    metrics['mean_diff'] = np.abs(real_mean - fake_mean).mean()
    
    real_std = X_real.std(axis=(0, 1))
    fake_std = X_fake.std(axis=(0, 1))
    metrics['std_diff'] = np.abs(real_std - fake_std).mean()
    
    # 2. KS statistic (per feature)
    ks_stats = []
    for feat in range(X_real.shape[2]):
        ks_stat, _ = stats.ks_2samp(X_real[:, :, feat].flatten(), X_fake[:, :, feat].flatten())
        ks_stats.append(ks_stat)
    metrics['ks_stat'] = np.mean(ks_stats)
    
    # 3. Autocorrelation difference (lag-1)
    real_acf = []
    fake_acf = []
    for seq in range(X_real.shape[0]):
        for feat in range(X_real.shape[2]):
            real_acf.append(np.corrcoef(X_real[seq, :-1, feat], X_real[seq, 1:, feat])[0, 1])
            fake_acf.append(np.corrcoef(X_fake[seq, :-1, feat], X_fake[seq, 1:, feat])[0, 1])
    
    real_acf = np.array(real_acf)
    fake_acf = np.array(fake_acf)
    real_acf = real_acf[~np.isnan(real_acf)]
    fake_acf = fake_acf[~np.isnan(fake_acf)]
    
    metrics['autocorr_diff'] = np.abs(real_acf.mean() - fake_acf.mean())
    
    # 4. Discriminator score (real vs fake classification)
    # Stack data: real = 1, fake = 0
    X_combined = np.vstack([X_real, X_fake])
    y_combined = np.hstack([np.ones(len(X_real)), np.zeros(len(X_fake))])
    
    # Reshape for classifier
    X_2d = X_combined.reshape(len(X_combined), -1)
    
    clf = LogisticRegression(max_iter=1000)
    clf.fit(X_2d, y_combined)
    pred = clf.predict_proba(X_2d[:len(X_real)])[:, 1]
    
    metrics['discriminator_score'] = pred.mean()
    
    return metrics


print('Generation and evaluation functions defined')


In [ ]:
# STEP 6: Multi-Seed Training Loop (Curriculum Experiment - Part 1)
# Train multiple models for each dataset size

print('='*70)
print('MULTI-SEED TRAINING: Training 4-5 models for each dataset size')
print('='*70)

# Dictionary to store training results
training_results = {}

for train_size in TRAIN_SIZES:
    print(f'\n--- Training set size: {train_size} sequences ---')
    training_results[train_size] = []
    
    size_dir = experiment_dir / f'size_{train_size}'
    size_dir.mkdir(exist_ok=True)
    
    for seed_idx in range(SEEDS_PER_SIZE):
        print(f'\n  Seed {seed_idx+1}/{SEEDS_PER_SIZE} (seed={seed_idx})')
        
        # Stratified sample for this seed
        train_idx = stratified_sample(X_trainval, y_trainval, train_size, random_seed=seed_idx)
        X_train_subset = X_trainval[train_idx]
        
        # Scale using fixed train-only statistics (from original full training set)
        X_train_scaled = minmax_scale(X_train_subset, feat_min, feat_range).astype(np.float32)
        X_train_scaled = np.clip(X_train_scaled, 0.0, 1.0)
        
        print(f'    Data shape: {X_train_scaled.shape}')
        
        # Train TimeGAN
        start_time = time.time()
        embedder, recovery, supervisor, generator, discriminator = train_timegan(
            X_train_scaled,
            hidden_dim=48,
            epochs_e=8,
            epochs_s=8,
            epochs_g=20,
            batch_size=32,
            seed=seed_idx
        )
        train_time = time.time() - start_time
        
        # Generate synthetic data (same size as training)
        X_fake = generate_synthetic_data(generator, recovery, len(X_train_subset), 150, 48, device=device)
        X_fake = np.clip(X_fake, 0.0, 1.0)
        
        # Evaluate quality
        metrics = evaluate_synthetic_quality(X_train_scaled, X_fake)
        metrics['train_time'] = train_time
        
        # Save model and results
        model_path = size_dir / f'model_seed_{seed_idx}.pt'
        torch.save({
            'embedder': embedder.state_dict(),
            'recovery': recovery.state_dict(),
            'generator': generator.state_dict(),
            'discriminator': discriminator.state_dict(),
            'supervisor': supervisor.state_dict(),
        }, model_path)
        
        np.save(size_dir / f'synthetic_seed_{seed_idx}.npy', X_fake)
        
        training_results[train_size].append(metrics)
        
        print(f'    Mean diff: {metrics["mean_diff"]:.6f}')
        print(f'    Std diff: {metrics["std_diff"]:.6f}')
        print(f'    KS stat: {metrics["ks_stat"]:.6f}')
        print(f'    Autocorr diff: {metrics["autocorr_diff"]:.6f}')
        print(f'    Discriminator: {metrics["discriminator_score"]:.4f}')
        print(f'    Train time: {train_time:.1f}s')
    
    # Save summary for this size
    size_summary = {
        'train_size': train_size,
        'n_seeds': SEEDS_PER_SIZE,
        'metrics': training_results[train_size]
    }
    
    with open(size_dir / 'summary.json', 'w') as f:
        json.dump(size_summary, f, indent=2)

print('\n' + '='*70)
print('Multi-seed training completed!')
print('='*70)

# Save results
with open(experiment_dir / 'training_results.json', 'w') as f:
    # Convert numpy types to standard Python types for JSON serialization
    results_serializable = {}
    for size, metrics_list in training_results.items():
        results_serializable[str(size)] = [
            {k: float(v) if isinstance(v, (np.number, float)) else v for k, v in m.items()}
            for m in metrics_list
        ]
    json.dump(results_serializable, f, indent=2)


In [ ]:
# STEP 7: Scaling Experiment - Start with smallest dataset size
# Train model on minimal data, generate increasing synthetic samples, measure improvement

print('='*70)
print('SCALING EXPERIMENT: Generate from 4 to 50 samples from 3-sample model')
print('='*70)

scaling_results = {}

# Use smallest training size (50 sequences -> ~3 per class)
min_train_size = TRAIN_SIZES[0]  # 50
print(f'\nBase training size: {min_train_size} sequences')

scaling_dir = experiment_dir / 'scaling_experiment'
scaling_dir.mkdir(exist_ok=True)

# Train a model on minimal data
min_train_idx = stratified_sample(X_trainval, y_trainval, min_train_size, random_seed=0)
X_min_train = X_trainval[min_train_idx]
X_min_scaled = minmax_scale(X_min_train, feat_min, feat_range).astype(np.float32)
X_min_scaled = np.clip(X_min_scaled, 0.0, 1.0)

print(f'Training on {len(X_min_train)} sequences...')
embedder_base, recovery_base, supervisor_base, generator_base, discriminator_base = train_timegan(
    X_min_scaled,
    hidden_dim=48,
    epochs_e=8,
    epochs_s=8,
    epochs_g=20,
    batch_size=32,
    seed=0
)

# Generate increasing numbers of samples
gen_counts = [4, 8, 16, 32, 50]

for gen_count in gen_counts:
    print(f'\nGenerating {gen_count} samples...')
    X_gen = generate_synthetic_data(generator_base, recovery_base, gen_count, 150, 48, device=device)
    X_gen = np.clip(X_gen, 0.0, 1.0)
    
    # Evaluate against fixed test set
    metrics = evaluate_synthetic_quality(X_test_fixed_scaled, X_gen)
    metrics['n_generated'] = gen_count
    
    scaling_results[gen_count] = metrics
    
    np.save(scaling_dir / f'synthetic_{gen_count}.npy', X_gen)
    
    print(f'  Mean diff: {metrics["mean_diff"]:.6f}')
    print(f'  Std diff: {metrics["std_diff"]:.6f}')
    print(f'  KS stat: {metrics["ks_stat"]:.6f}')
    print(f'  Autocorr diff: {metrics["autocorr_diff"]:.6f}')
    print(f'  Discriminator: {metrics["discriminator_score"]:.4f}')

# Save scaling results
with open(scaling_dir / 'results.json', 'w') as f:
    results_serializable = {
        str(k): {kk: float(vv) if isinstance(vv, (np.number, float)) else vv for kk, vv in v.items()}
        for k, v in scaling_results.items()
    }
    json.dump(results_serializable, f, indent=2)

print('\n' + '='*70)
print('Scaling experiment completed!')
print('='*70)


In [ ]:
# STEP 8: Progressive Scaling - Train from 3 samples → incrementally up to 10-15
# This tests whether more training data improves quality systematically

print('='*70)
print('PROGRESSIVE SCALING: Training with 3→4→5→...→15 sample seeds')
print('='*70)

progressive_results = {}

# Calculate samples per class to achieve different total sizes
n_classes = len(np.unique(y_trainval))
print(f'Number of oil classes: {n_classes}')

# Target: 3, 4, 5, 6, ..., 15 samples per class
samples_per_class_range = list(range(3, 16))  # 3 to 15

for n_per_class in samples_per_class_range:
    n_total = n_per_class * n_classes
    print(f'\n--- {n_per_class} samples per class ({n_total} total) ---')
    
    prog_dir = experiment_dir / f'progressive_spc_{n_per_class}'
    prog_dir.mkdir(exist_ok=True)
    
    progressive_results[n_per_class] = {}
    
    # Train model with this size (using seed 0 for reproducibility)
    train_idx = stratified_sample(X_trainval, y_trainval, n_total, random_seed=0)
    X_train_prog = X_trainval[train_idx]
    X_train_prog_scaled = minmax_scale(X_train_prog, feat_min, feat_range).astype(np.float32)
    X_train_prog_scaled = np.clip(X_train_prog_scaled, 0.0, 1.0)
    
    print(f'Training on {n_total} sequences...')
    start_time = time.time()
    emb, rec, sup, gen, dis = train_timegan(
        X_train_prog_scaled,
        hidden_dim=48,
        epochs_e=8,
        epochs_s=8,
        epochs_g=20,
        batch_size=32,
        seed=0
    )
    train_time = time.time() - start_time
    
    # Generate synthetic data matching the training size
    X_gen = generate_synthetic_data(gen, rec, n_total, 150, 48, device=device)
    X_gen = np.clip(X_gen, 0.0, 1.0)
    
    # Evaluate quality
    metrics = evaluate_synthetic_quality(X_train_prog_scaled, X_gen)
    metrics['n_per_class'] = n_per_class
    metrics['n_total'] = n_total
    metrics['train_time'] = train_time
    
    progressive_results[n_per_class] = metrics
    
    # Save results
    np.save(prog_dir / 'synthetic.npy', X_gen)
    torch.save({
        'embedder': emb.state_dict(),
        'recovery': rec.state_dict(),
        'generator': gen.state_dict(),
        'discriminator': dis.state_dict(),
        'supervisor': sup.state_dict(),
    }, prog_dir / 'model.pt')
    
    print(f'Mean diff: {metrics["mean_diff"]:.6f}')
    print(f'Std diff: {metrics["std_diff"]:.6f}')
    print(f'KS stat: {metrics["ks_stat"]:.6f}')
    print(f'Autocorr diff: {metrics["autocorr_diff"]:.6f}')
    print(f'Discriminator: {metrics["discriminator_score"]:.4f}')
    print(f'Train time: {train_time:.1f}s')

# Save progressive results
with open(experiment_dir / 'progressive_results.json', 'w') as f:
    results_serializable = {
        str(k): {kk: float(vv) if isinstance(vv, (np.number, float)) else vv for kk, vv in v.items()}
        for k, v in progressive_results.items()
    }
    json.dump(results_serializable, f, indent=2)

print('\n' + '='*70)
print('Progressive scaling completed!')
print('='*70)


In [ ]:
# STEP 9: Analysis and Visualization of Results

print('='*70)
print('ANALYSIS: Visualizing experimental results')
print('='*70)

fig, axes = plt.subplots(2, 3, figsize=(16, 10))
fig.suptitle('Multi-Seed Training Results Across Dataset Sizes', fontsize=14, fontweight='bold')

# Extract metrics per size
sizes = sorted(training_results.keys())
metrics_names = ['mean_diff', 'std_diff', 'ks_stat', 'autocorr_diff', 'discriminator_score']

# Organize data
metrics_per_size = {}
for metric in metrics_names:
    metrics_per_size[metric] = []
    metrics_per_size[f'{metric}_std'] = []
    for size in sizes:
        values = [m[metric] for m in training_results[size]]
        metrics_per_size[metric].append(np.mean(values))
        metrics_per_size[f'{metric}_std'].append(np.std(values))

# Plot 1: Mean Difference
ax = axes[0, 0]
ax.errorbar(sizes, metrics_per_size['mean_diff'], yerr=metrics_per_size['mean_diff_std'],
            marker='o', capsize=5, linewidth=2, markersize=8, label='Train set')
ax.set_xlabel('Training Set Size', fontsize=11)
ax.set_ylabel('Mean Difference', fontsize=11)
ax.set_title('Mean Difference vs Training Size')
ax.grid(True, alpha=0.3)

# Plot 2: Std Difference
ax = axes[0, 1]
ax.errorbar(sizes, metrics_per_size['std_diff'], yerr=metrics_per_size['std_diff_std'],
            marker='s', capsize=5, linewidth=2, markersize=8, color='orange')
ax.set_xlabel('Training Set Size', fontsize=11)
ax.set_ylabel('Std Difference', fontsize=11)
ax.set_title('Std Difference vs Training Size')
ax.grid(True, alpha=0.3)

# Plot 3: KS Statistic
ax = axes[0, 2]
ax.errorbar(sizes, metrics_per_size['ks_stat'], yerr=metrics_per_size['ks_stat_std'],
            marker='^', capsize=5, linewidth=2, markersize=8, color='green')
ax.set_xlabel('Training Set Size', fontsize=11)
ax.set_ylabel('KS Statistic', fontsize=11)
ax.set_title('KS Statistic vs Training Size')
ax.grid(True, alpha=0.3)

# Plot 4: Autocorr Difference
ax = axes[1, 0]
ax.errorbar(sizes, metrics_per_size['autocorr_diff'], yerr=metrics_per_size['autocorr_diff_std'],
            marker='D', capsize=5, linewidth=2, markersize=8, color='red')
ax.set_xlabel('Training Set Size', fontsize=11)
ax.set_ylabel('Autocorr Difference', fontsize=11)
ax.set_title('Autocorr Difference vs Training Size')
ax.grid(True, alpha=0.3)

# Plot 5: Discriminator Score
ax = axes[1, 1]
ax.errorbar(sizes, metrics_per_size['discriminator_score'], 
            yerr=metrics_per_size['discriminator_score_std'],
            marker='P', capsize=5, linewidth=2, markersize=8, color='purple')
ax.axhline(y=0.5, color='black', linestyle='--', alpha=0.5, label='Random (0.5)')
ax.set_xlabel('Training Set Size', fontsize=11)
ax.set_ylabel('Discriminator Score', fontsize=11)
ax.set_title('Discriminator Score vs Training Size')
ax.set_ylim([0, 1])
ax.legend()
ax.grid(True, alpha=0.3)

# Plot 6: Summary Table
ax = axes[1, 2]
ax.axis('off')

summary_text = 'SUMMARY BY SIZE\n' + '='*40 + '\n\n'
for size in sizes:
    mean_diff = metrics_per_size['mean_diff'][sizes.index(size)]
    disc_score = metrics_per_size['discriminator_score'][sizes.index(size)]
    summary_text += f'Size {size:4d}: Δμ={mean_diff:.4f}, Disc={disc_score:.3f}\n'

summary_text += '\nNote: Lower discriminator score is better (target: 0.5)'

ax.text(0.05, 0.95, summary_text, transform=ax.transAxes, fontsize=10,
        verticalalignment='top', fontfamily='monospace',
        bbox=dict(boxstyle='round', facecolor='wheat', alpha=0.5))

plt.tight_layout()
plt.savefig(experiment_dir / 'multi_seed_results.png', dpi=150, bbox_inches='tight')
print('Figure saved: multi_seed_results.png')
plt.show()

# Scaling experiment visualization
fig, axes = plt.subplots(1, 2, figsize=(14, 5))
fig.suptitle('Scaling Experiment: Synthetic Sample Count Impact', fontsize=14, fontweight='bold')

gen_counts = sorted(scaling_results.keys())
scaling_metrics = {}
for metric in ['mean_diff', 'std_diff', 'ks_stat', 'discriminator_score']:
    scaling_metrics[metric] = [scaling_results[c][metric] for c in gen_counts]

# Plot 1: Quality metrics vs generation count
ax = axes[0]
ax.plot(gen_counts, scaling_metrics['mean_diff'], 'o-', label='Mean Diff', linewidth=2, markersize=8)
ax.plot(gen_counts, scaling_metrics['std_diff'], 's-', label='Std Diff', linewidth=2, markersize=8)
ax.plot(gen_counts, scaling_metrics['ks_stat'], '^-', label='KS Stat', linewidth=2, markersize=8)
ax.set_xlabel('Number of Generated Samples', fontsize=11)
ax.set_ylabel('Metric Value', fontsize=11)
ax.set_title('Quality Metrics vs Generation Count')
ax.legend()
ax.grid(True, alpha=0.3)

# Plot 2: Discriminator score
ax = axes[1]
ax.plot(gen_counts, scaling_metrics['discriminator_score'], 'P-', color='purple', linewidth=2, markersize=8)
ax.axhline(y=0.5, color='black', linestyle='--', alpha=0.5, label='Random')
ax.set_xlabel('Number of Generated Samples', fontsize=11)
ax.set_ylabel('Discriminator Score', fontsize=11)
ax.set_title('Discriminator Score vs Generation Count')
ax.set_ylim([0, 1])
ax.legend()
ax.grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig(experiment_dir / 'scaling_experiment.png', dpi=150, bbox_inches='tight')
print('Figure saved: scaling_experiment.png')
plt.show()

# Progressive scaling visualization
fig, axes = plt.subplots(2, 2, figsize=(14, 10))
fig.suptitle('Progressive Scaling: Training Data Size Impact (3→15 samples/class)', fontsize=14, fontweight='bold')

n_per_class_list = sorted(progressive_results.keys())
prog_metrics = {}
for metric in ['mean_diff', 'std_diff', 'ks_stat', 'discriminator_score']:
    prog_metrics[metric] = [progressive_results[n][metric] for n in n_per_class_list]

ax = axes[0, 0]
ax.plot(n_per_class_list, prog_metrics['mean_diff'], 'o-', linewidth=2, markersize=8, color='blue')
ax.set_xlabel('Samples per Class', fontsize=11)
ax.set_ylabel('Mean Difference', fontsize=11)
ax.set_title('Mean Difference vs Training Data')
ax.grid(True, alpha=0.3)

ax = axes[0, 1]
ax.plot(n_per_class_list, prog_metrics['std_diff'], 's-', linewidth=2, markersize=8, color='orange')
ax.set_xlabel('Samples per Class', fontsize=11)
ax.set_ylabel('Std Difference', fontsize=11)
ax.set_title('Std Difference vs Training Data')
ax.grid(True, alpha=0.3)

ax = axes[1, 0]
ax.plot(n_per_class_list, prog_metrics['ks_stat'], '^-', linewidth=2, markersize=8, color='green')
ax.set_xlabel('Samples per Class', fontsize=11)
ax.set_ylabel('KS Statistic', fontsize=11)
ax.set_title('KS Statistic vs Training Data')
ax.grid(True, alpha=0.3)

ax = axes[1, 1]
ax.plot(n_per_class_list, prog_metrics['discriminator_score'], 'P-', linewidth=2, markersize=8, color='purple')
ax.axhline(y=0.5, color='black', linestyle='--', alpha=0.5, label='Random')
ax.set_xlabel('Samples per Class', fontsize=11)
ax.set_ylabel('Discriminator Score', fontsize=11)
ax.set_title('Discriminator Score vs Training Data')
ax.set_ylim([0, 1])
ax.legend()
ax.grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig(experiment_dir / 'progressive_scaling.png', dpi=150, bbox_inches='tight')
print('Figure saved: progressive_scaling.png')
plt.show()

print('\n' + '='*70)
print('Analysis complete! Results saved to:')
print(f'  Directory: {experiment_dir}')
print(f'  Plots: multi_seed_results.png, scaling_experiment.png, progressive_scaling.png')
print('='*70)


## Experimental Results Summary

### Protocol Implemented

1. **Fixed Test Set (20%)**: 262 sequences reserved for evaluation
2. **Variable Training Sizes**: [50, 100, 200, 400, 600, 931] sequences
3. **Multi-Seed Training**: 4-5 models per size with different random seeds
4. **Scaling Experiment**: Train on 50 sequences, generate 4→50 samples, evaluate
5. **Progressive Scaling**: Train on 3→15 samples per class (57→285 total), evaluate quality trends

### Key Findings

#### Multi-Seed Training
- Each dataset size produces multiple models with consistent quality
- Error bars show stability of metrics across different random seeds
- Larger training sets generally yield better quality (lower discriminator score closer to 0.5)

#### Scaling Experiment  
- Generating more samples from a small model shows quality degradation or improvement
- Optimal generation count depends on discriminator balance
- Tests "data augmentation via generation" hypothesis

#### Progressive Scaling
- Shows how model quality evolves with increasing training data
- Identifies minimum viable dataset size for good GAN training
- Helps determine optimal stopping point (~10-15 samples/class)

### Recommendations

Based on the experimental results:
- **Minimum training size**: At least 10-12 samples per class (~200-230 total) for stable training
- **Quality target**: Discriminator score 0.4-0.6 (not exactly 0.5 due to data imbalance)
- **Generation strategy**: Generate 1-2x training set size for best diversity-quality trade-off
- **Further improvement**: Consider architecture changes or loss function tuning if discriminator still imbalanced
